File: go/leet_code/42.go
package leet_code

// trap calculates the total amount of water that can be trapped between the bars.
// The heights of the bars are given by the `height` array.
// It uses two pointers, `left` and `right`, starting from the ends of the array and moving towards the center.
// `leftWall` and `rightWall` keep track of the highest bar from the left and right, respectively.
// The water trapped at a certain index is the difference between the height of the wall and the height of the bar at that index.
func trap(height []int) int {
	left, right := 0, len(height)-1 // Initialize pointers at both ends of the array.
	leftWall, rightWall := 0, 0     // Initialize the left and right walls with height 0.
	ans := 0                        // Initialize the answer, which is the total amount of trapped water.

	// Move the pointers towards the center.
	for left < right {
		// If the bar at the left pointer is shorter than the bar at the right pointer,
		// update the left wall and add the trapped water at the left index to the answer.
		if height[left] < height[right] {
			leftWall = max(height[left], leftWall)
			ans += leftWall - height[left]
			left++
		} else {
			// If the bar at the right pointer is shorter or equal to the bar at the left pointer,
			// update the right wall and add the trapped water at the right index to the answer.
			rightWall = max(height[right], rightWall)
			ans += rightWall - height[right]
			right--
		}
	}

	// Return the total amount of trapped water.
	return ans
}

File: rust/src/leet_code/42.rs
/// This module provides a solution for the trapping rainwater problem.
///
/// It includes a function `trap` that calculates the amount of rainwater that can be trapped
/// given a list of heights.
use std::cmp;

impl Solution {
    /// Calculate the amount of rainwater that can be trapped given a list of heights.
    ///
    /// # Arguments
    ///
    /// * `height` - A vector of i32s representing the heights.
    ///
    /// # Returns
    ///
    /// * An i32 representing the amount of trapped rainwater.
    pub fn trap(height: Vec<i32>) -> i32 {
        // Initialize peak index
        let mut peak_i = 0;
        let mut i = 1;
        // Find the peak of the height
        loop {
            if i == height.len() {
                break;
            }
            if height[i] > height[peak_i] {
                peak_i = i;
            }
            i += 1;
        }

        // Initialize water and left_max
        let mut water = 0;
        let mut i = 0;
        let mut left_max = height[0];
        // Calculate water trapped on the left side of the peak
        loop {
            if i == peak_i {
                break;
            }
            if height[i] > left_max {
                left_max = height[i];
            }
            water += left_max - height[i];
            i += 1;
        }

        // Initialize right_max
        let mut i = height.len() - 1;
        let mut right_max = height[i];
        // Calculate water trapped on the right side of the peak
        loop {
            if i == peak_i {
                return water;
            }
            if height[i] > right_max {
                right_max = height[i];
            }
            water += right_max - height[i];
            i -= 1;
        }
    }
}
